cmd/snap/cmd_run.go

```diff
@@ -216,6 +220,72 @@ func snapRunHook(snapName, snapRevision, hookName string) error {
 	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, nil)
 }
 
+func migrateXauthority(info *snap.Info) error {
+	u, err := userCurrent()
+	if err != nil {
+		return fmt.Errorf(i18n.G("cannot get the current user: %v"), err)
+	}
+
+	// If we're running as root then we don't do anything
+	if u.Uid == "0" {
+		return nil
+	}
+
+	xauthPath := osGetenv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return nil
+	}
+
+	// Copy Xauthority file into a temporary place so we can safely
+	// process it further there.
+	tmpDir, err := ioutil.TempDir("", "xauth")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+	flags := osutil.CopyFlagSync | osutil.CopyFlagPreserveAll | osutil.CopyFlagOverwrite
+	tmpXauthPath := filepath.Join(tmpDir, "Xauthority")
+	err = osutil.CopyFile(xauthPath, tmpXauthPath, flags)
+	if err != nil {
+		return err
+	}
+
+	baseTargetDir := filepath.Join(dirs.XdgRuntimeDirBase, u.Uid)
+	targetPath := filepath.Join(baseTargetDir, "Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// ohterwise we can continue using the existing Xauthority file
+	if osutil.FileExists(targetPath) && osutil.FilesAreEqual(targetPath, tmpXauthPath) {
+		osSetenv("XAUTHORITY", targetPath)
+		return nil
+	}
+
+	// Ensure that we have a valid Xauthority. It is invalid when
+	// either the data can't be parsed or there are no cookies in
+	// the file is invalid.
+	if err := x11.ValidateXauthority(tmpXauthPath); err != nil {
+		return nil
+	}
+
+	if !osutil.FileExists(baseTargetDir) {
+		if err := os.MkdirAll(baseTargetDir, 0700); err != nil {
+			return err
+		}
+	}
+
+	if err = osutil.CopyFile(tmpXauthPath, targetPath, flags); err != nil {
+		return err
+	}
+
+	// If everything is ok, we can now point the snap to the new
+	// location of the Xauthority file.
+	osSetenv("XAUTHORITY", targetPath)
```

Comment:
Just re-stating the `setenv`-is-dangerous comment I made earlier. I'll let @mvo5 comment on this, if we want to re-work it to avoid setenv entirely.